Delayed nu fission score and delayed group filter

samuelshaner opened the pull request:
This PR incorporates two features into OpenMC necessary to compute delayed fractions and delayed neutron spectrums for the different delayed groups. The features include:

- Delayed nu fission score `delay-nu-fission`: This score can be used to tally the delayed fission production rate, either cumulatively or in each delayed group separately. Energy in and energy out filters are supported, which are necessary to compute the delayed neutron spectrums.
- Delayed group filter `delaygroup`: This filter can only be used with the `delay-nu-fission` score and allows tallies of the delayed fission rate to be formed for all delayed groups. There are no bin options for this filter; rather the bins attribute is set to an empty string and the delayed fission rate is tallied for each delayed group separately. Setting the bins attribute to an empty string seems awkward, so I'm open to thoughts on other ways to incorporate this filter.

One complication of implementing these features is the different number of delayed groups in different cross section libraries (e.g. ENDF/B-VII.1 has 6 and JEFF 3.1.1 has 8 delayed groups). Since the number of delayed groups is not known prior to reading in the input files, a constant `MAX_DELAYED_GROUPS` is set to 8 so the `delay-nu-fission` score bins can be used for the 6 and 8 delayed group structures. The `delay-nu-fission` score is then output for the `MAX_DELAYED_GROUPS`, with any groups not present returned as 0.0. A check is performed when the ace files are parsed and an error message is output if a nuclide with a number of delayed groups greater than `MAX_DELAYED_GROUPS` is encountered.

paulromano commented:
Thanks for the pull request @samuelshaner. It looks like the regression tests are failing when HDF5 is enabled. Can you look into this? Once all tests pass, I'll perform a full review so we can get this merged in.

samuelshaner commented:
Sure, I'll look into this and update my PR once I have all the tests passing.

nelsonag commented:
One note of caution: as you can imagine the delayed neutron groups are not consistent across fissionable isotopes.  That is, group 1 for precursors from U-235 has a different average half-life than does group 1 for precursors from U-238.  So technically implementing this in anything but nuclide-specific tallies could be misleading to the users.

samuelshaner commented:
@nelsonag that's a good point. The ENDF/B-VII.1 6 group delayed neutron data has different half-lives for some delayed groups across the set of nuclides. I believe the JEFF 3.1.1 8 group delayed neutron data has the same half-life for all nuclides for each delayed group, so nuclide-specific tallies would not be required. I'll look into adding a warning for when the `delaygroup` filter is used in combination of 6 group data and I'll remember to add commentary on this topic in the documentation.

samuelshaner commented:
I have updated this PR with the following modifications/additions:
- The `delaygroup` filter and `delay-nu-fission` score have been changed to `delayedgroup` and `delayed-nu-fission`, respectively.
- Added a warning if the `delayedgroup` filter is used with a total nuclide tally. Since the delayed group structure is not the same for all isotopes in ENDF but is the same for all isotopes in JEFF, users should be prompted with this warning.
- The `delayedgroup` filter now has the option for the user to set the delayed groups they want to tally to in.
- Added documentation for the `delayedgroup` filter and `delayed-nu-fission` score.
- The material `delayed_nu_fission` attribute is now an array of cross sections for each delayed group.

@paulromano - this should be ready for your review.

paulromano commented:
@samuelshaner Thanks for updating this -- I was wondering what the status was but it's good to see you still intend to have it merged. With that being said, there are a few things you still need: 
- Tests that use the new score/filter
- Modify the Python API so that it knows about the score/filter
- Modify RELAX NG schema so that the new filter is considered valid input

You might want to take a peak at #470 to get an idea of where you'll need to make changes. Always easier to follow in someone else's footsteps rather than just blindly thumb through the source figuring out where changes are needed!

Once everything is there, I'll begin a full review.

samuelshaner commented:
@paulromano - thanks for the useful comments. I have addressed the issues you brought up and it should be ready for review.

paulromano commented on docs/source/usersguide/input.rst:
```diff
@@ -1276,7 +1276,13 @@ The ``<tally>`` element accepts the following sub-elements:
+          A list of delayed neutron precursor groups for which the tally should
+          be accumulated. For instance, to tally to all 6 delayed groups in the
+          ENDF/B-VII.1 library the filter is specified as ``<filter
+          type="delayedgroup" bins="1 2 3 4 5 6" />``.
```
Use `.. code-block:: xml` for the example

paulromano commented on src/initialize.F90:
```diff
@@ -223,18 +223,19 @@ subroutine initialize_mpi()
-    bank_types = (/ MPI_REAL8, MPI_REAL8, MPI_REAL8, MPI_REAL8 /)
-    call MPI_TYPE_CREATE_STRUCT(4, bank_blocks, bank_disp, &
+    bank_blocks = (/ 1, 3, 3, 1, 1 /)
+    bank_types = (/ MPI_REAL8, MPI_REAL8, MPI_REAL8, MPI_REAL8, MPI_REAL8 /)
```
Last one should be `MPI_INTEGER`

paulromano commented on src/bank_header.F90:
```diff
@@ -11,10 +11,11 @@ module bank_header
+    real(C_DOUBLE) :: xyz(3)        ! location of bank particle
+    real(C_DOUBLE) :: uvw(3)        ! diretional cosines
+    real(C_DOUBLE) :: E             ! energy
+    integer        :: delayed_group ! delayed group
```
Type this as `integer(C_INT)`

paulromano commented on src/cross_section.F90:
```diff
@@ -46,6 +49,10 @@ subroutine calculate_xs(p)
 
+    do d = 1, MAX_DELAYED_GROUPS
+      material_xs % delayed_nu_fission(d) = ZERO
+    end do
```
`material_xs % delayed_nu_fission(:) = ZERO` eliminates the do loop

paulromano commented on src/cross_section.F90:
```diff
@@ -133,6 +142,15 @@ subroutine calculate_xs(p)
+        material_xs % delayed_nu_fission(d) = &
+             material_xs % delayed_nu_fission(d) +&
+             atom_density * micro_xs(i_nuclide) % delayed_nu_fission * yield
+      end do
```
Unfortunately this block of code here is a showstopper. MC code performance is particularly sensitive to cross section lookups, and this code makes that lookup quite a bit more onerous. I just ran a simple test on the Hoogenboom-Martin benchmark and the performance drops by more than 50%.

My suggestion is to calculate these micro and macroscopic cross sections on-the-fly when needed rather than calculating them up-front here every time the particle's energy changes and storing them in the micro/macro caches. Otherwise, this hurts performance pretty severely for a simulation that doesn't even use your new score/filter.

samuelshaner replied:
Okay, I'll make functions to compute the micro/macro delayed-nu-fission cross sections.

paulromano commented on src/cross_section.F90:
```diff
@@ -248,6 +267,12 @@ subroutine calculate_nuclide_xs(i_nuclide, i_sab, E, i_mat, i_nuc_mat, u)
+      ! Calculate microscopic nuclide delayed nu-fission cross section
+      micro_xs(i_nuclide) % delayed_nu_fission = (ONE - f) * &
+           nuc % delayed_nu_fission(i_grid) + f * &
+           nuc % delayed_nu_fission(i_grid+1)
```
This is also related to the performance drop. I haven't profiled anything, so I don't actually know which of these is more expensive. It may be that calculating these microscopic cross sections is more costly. Keep in mind the cost really has nothing to do with the meager amount of FLOPs needed but rather the fact that going to fetch data from `nuc%delayed_nu_fission` likely means missing cache and having to go out to main memory.

samuelshaner replied:
Removing the macroscopic `delayed_nu_fission` xs calculation in `calculate_xs` is not a big deal. The macro xs is only needed when performing a track length or collision tally on the total nuclides, and the tally routine can just be modified to loop over all nuclides in a material and tally the contributions from each nuclide. Removing the microscopic `delayed_nu_fission` xs from the `calculate_nuclide_xs` subroutine would require a new subroutine to calculate the nuclide delayed-nu-fission microscopic xs. While this is not difficult to implement, it results in the cross sections not being completely in sync with each other. When a delayed-nu-fission score is needed, the nuclide delayed-nu-fission microscopic xs would need to be recomputed, which could be confusing to require this dependency for only one tally score.

Would it be okay to keep in the delayed-nu-fission micro xs calculation, but remove the delayed-nu-fission macro xs calculation? Is there any profiling you would recommend doing to investigate the performance drop of calculating the `delayed-nu-fission` microscopic xs in the `calculate_xs` subroutine?

paulromano replied:
There are other tally scores where microscopic cross sections are calculated as needed (just look at the case default in tally.F90:score_general). It's hard to say if it's ok to keep the delayed-nu-fission micro calculation without knowing the impact. For profiling, you have two options -- you can take the lazy approach of just picking a reactor problem and seeing what the difference in performance is (i.e. no profiling really). You can also add profiling flags and use gprof to get an actual profile:
```sh
cmake -Dprofile=on ..
make
...
openmc
gprof openmc_executable gmon.out > profile.txt
```

paulromano replied:
If you're using Intel compilers, they also have much fancier tools for profiling.

samuelshaner replied:
I used gprof to profile the code with and without the computation of the microscopic delayed-nu-fission cross section in the `calculate_nuclide_xs` subroutine. I ran the `examples/xml/lattice/simple` case with 1e6 particles/batch three times with both versions of the code. The `calculate_nuclide_xs` subroutine took up ~4-5% of the total run time for both code versions. In the cases that include the calculation of the microscopic delayed-nu-fission xs, the run times for the `calculate_nuclide_xs` subroutine were 10.03, 8.88, and 9.77 seconds for an average of 9.5 seconds. In the cases that did not include the calculation of the microscopic delayed-nu-fission xs, the run times for the `calculate_nuclide_xs` subroutine were 9.84, 9.12, and 9.00 seconds for an average of 9.3 seconds. All cases were run on an AWS m4.10xlarge instance with one MPI thread. This is clearly a small sample size, but nevertheless it seems to indicate that the calculation of the microscopic delayed-nu-fission xs does not have a significant impact on run time.

Is there a specific input you would recommend I run in profile mode to get a more representative indication for the impact of adding the calculation of the microscopic delayed-nu-fission xs?

paulromano replied:
The problem with that model is that it only includes 3 nuclides, so it's not surprising there is not much effect because there aren't very many cross section lookups to start with. I use the [Monte Carlo performance benchmark](https://github.com/mit-crpg/benchmarks/tree/master/mc-performance/openmc) (also known as Hoogenboom-Martin) often as a quick check (which has ~40-50 nuclides), but BEAVRS is a more appropriate real problem, if you have the model.

samuelshaner replied:
Ok, I'll test out Hoogenboom-Martin and, if the results still look reasonable, try out BEAVRS.

paulromano commented on src/initialize.F90:
```diff
@@ -306,6 +307,8 @@ subroutine hdf5_initialize()
     call h5tinsert_f(hdf5_bank_t, "E", h5offsetof(c_loc(tmpb(1)), &
          c_loc(tmpb(1)%E)), H5T_NATIVE_DOUBLE, hdf5_err)
+    call h5tinsert_f(hdf5_bank_t, "delayed_group", h5offsetof(c_loc(tmpb(1)), &
+         c_loc(tmpb(1)%delayed_group)), H5T_NATIVE_DOUBLE, hdf5_err)
```
Second-to-last argument should be `H5T_NATIVE_INTEGER`

paulromano commented on src/input_xml.F90:
```diff
@@ -2568,6 +2569,28 @@ subroutine read_tallies_xml()
+                call fatal_error("Encountered delayedgroup bin with index " &
+                     &// trim(to_str(t % filters(j) % int_bins(d))) &
+                     &//" that is outside the range of 1 to MAX_DELAYED_GROUPS"&
+                     &// " (" // trim(to_str(MAX_DELAYED_GROUPS)) // ")")
```
Ampersand beginning a continuation line is unnecessary (only needed if you are in the middle of a string), e.g.
```
string = "This is a very very very &
     &long sentence."
```

paulromano commented on src/input_xml.F90:
```diff
@@ -2907,26 +3039,63 @@ subroutine read_tallies_xml()
+            if (t % find_filter(FILTER_DELAYEDGROUP) > 0) then
+              call fatal_error("Cannot tally current with a &
+                   &delayedgroup energy filter.")
+            end if
```
A lot of redundant code here. Why not just have a check after the select case with something like,
```Fortran
if (score_name /= 'delayed-nu-fission' .and. t%find_filter(FILTER_DELAYEDGROUP) > 0) then
  ...
end if
```

paulromano commented on src/fission.F90:
```diff
@@ -102,4 +106,57 @@ function nu_delayed(nuc, E) result(nu)
+
+  function yield_delayed(nuc, E, g) result(yield)
+
+    type(Nuclide), pointer :: nuc   ! nuclide from which to find nu
```
Use `type(Nuclide), intent(in)`

paulromano commented on src/fission.F90:
```diff
@@ -102,4 +106,57 @@ function nu_delayed(nuc, E) result(nu)
+    type(Nuclide), pointer :: nuc   ! nuclide from which to find nu
+    real(8), intent(in)    :: E     ! energy of incoming neutron
+    real(8)                :: yield ! delayed neutron precursor yield
+    integer                :: g     ! the delayed neutron precursor group
```
Declare `intent(in)`

paulromano commented on src/fission.F90:
```diff
@@ -102,4 +106,57 @@ function nu_delayed(nuc, E) result(nu)
+! a given nuclide and incoming neutron energy in a given delayed group.
+!===============================================================================
+
+  function yield_delayed(nuc, E, g) result(yield)
```
Make this function `pure`

samuelshaner replied:
This can only be made a `pure` function if `interpolate_tab1` is also made `pure`. Then `interpolate_tab1` can only be made pure if `binary_search` and `fatal_error` are made `pure`. It seems like all the functions in `fission.F90` should be made pure, but doing so requires making several other functions in other input files to be `pure`. How about we just make an issue to go through and purify all qualifying functions?

paulromano replied:
Good point -- yes we should just handle the pure issues separately.

paulromano commented on src/particle_header.F90:
```diff
@@ -122,6 +127,11 @@ subroutine initialize_particle(this)
+
+    do d = 1, MAX_DELAYED_GROUPS
+      this % n_delayed_bank(d) = 0
+    end do
```
`this % n_delayed_bank(:) = 0`

paulromano commented on src/physics.F90:
```diff
@@ -1108,6 +1110,13 @@ subroutine create_fission_sites(p, i_nuclide, i_reaction)
+    ! to zero.
+    do d = 1, MAX_DELAYED_GROUPS
+      nu_delayed(d) = 0
+    end do
```
`nu_delayed(:) = 0`

paulromano commented on src/physics.F90:
```diff
@@ -1130,15 +1139,26 @@ subroutine create_fission_sites(p, i_nuclide, i_reaction)
     p % wgt_bank = nu/weight
+    do d = 1, MAX_DELAYED_GROUPS
+      p % n_delayed_bank(d) = nu_delayed(d)
+    end do
```
`p % n_delayed_bank(:) = nu_delayed(:)`

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +977,124 @@ subroutine score_fission_eout(p, t, i_score)
+  subroutine score_fission_delayed_eout(p, t, i_score)
+
+    type(Particle), intent(in) :: p
+    type(TallyObject), pointer :: t
```
`type(TallyObject), intent(in)`

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +977,124 @@ subroutine score_fission_eout(p, t, i_score)
+
+  subroutine score_fission_delayed_dg(t, d_bin, score, score_index)
+
+    type(TallyObject), pointer :: t
```
`type(TallyObject), intent(in)`

paulromano commented on src/tracking.F90:
```diff
@@ -164,6 +166,10 @@ subroutine transport(p)
 
+        do d = 1, MAX_DELAYED_GROUPS
+          p % n_delayed_bank = 0
+        end do
```
`p % n_delayed_bank(:) = 0`

samuelshaner commented:
@paulromano - I have addressed your comments and made the following modifications:

- Removed the computation of the macroscopic and microscopic `delayed-nu-fission` cross sections. When the `delayed-nu-fission` cross section is needed for a particular delayed group, it is computed using the fission cross section and `nu-delayed`, which is computed on-the-fly.
- The total nuclide tallies have been modified to loop over all nuclides in the material.
- Comments have been added to the `delayed-nu-fission` score in `tally.F90` to better describe the process for each different case.

paulromano commented on docs/source/usersguide/input.rst:
```diff
@@ -1346,7 +1346,16 @@ The ``<tally>`` element accepts the following sub-elements:
+
+              <filter type="delayedgroup" bins="1 2 3 4 5 6" />
+
+:nuclides:
```
Looks like indentation for this line got changed accidentally

paulromano commented on openmc/statepoint.py:
```diff
@@ -430,13 +430,12 @@ def tallies(self):
-                        score = score.replace('-yn', '-' + moments[j].decode())
+
+                    # If this is a moment, use generic moment order
+                    regexp = re.compile(r'-n$|-pn$|-yn$')
```
You should either compile the regex when the module is loaded, or just use the `re.search` function instead (i.e. don't compile ahead of time).

paulromano commented on openmc/statepoint.py:
```diff
@@ -430,13 +430,12 @@ def tallies(self):
+                    regexp = re.compile(r'-n$|-pn$|-yn$')
+                    if regexp.search(score) is not None:
+                        score = score.strip(regexp.findall(score)[0])
+                        score += '-' + moments[j].decode()
```
These two lines could be replaced with `re.sub` or `regexp.sub`, depending on which approach you take.

paulromano commented on src/ace.F90:
```diff
@@ -635,6 +635,15 @@ subroutine read_nu_data(nuc)
+      ! of delayed groups
+      if (NPCR > MAX_DELAYED_GROUPS) then
+        call fatal_error("Encountered nuclide with " // trim(to_str(NPCR)) &
+             &// " delayed groups while the maximum number of delayed groups " &
```
Starting ampersand is redundant on these continuation lines

paulromano commented on src/input_xml.F90:
```diff
@@ -2884,6 +2935,14 @@ subroutine read_tallies_xml()
 
+          ! Check if delayed group filter is used with any score besides
+          ! delayed-nu-fission
+          if (trim(score_name) /= 'delayed-nu-fission' .and. &
```
`score_name /= 'delayed-nu-fission'` is sufficient

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1041,122 @@ subroutine score_fission_eout(p, t, i_score)
+
+  subroutine score_fission_delayed_dg(t, d_bin, score, score_index)
+
+    type(TallyObject)   :: t
```
need `intent(inout)`

paulromano commented on src/tally.F90:
```diff
@@ -395,6 +402,195 @@ subroutine score_general(p, t, start_index, filter_index, i_nuclide, &
+              ! from each group
+              do d = 1, nuclides(p % event_nuclide) % n_precursor
+                score = score + keff * p % wgt_bank / p % n_bank * &
+                     p % n_delayed_bank(d)
```
The factors `keff * p % wgt_bank / p % n_bank` are constant and thus don't really need to be in the loop. A simpler way might be:
```fortran
score = keff * p % wgt_bank / p % n_bank * &
     sum(p % n_delayed_bank)
```

paulromano commented on src/tally.F90:
```diff
@@ -395,6 +402,195 @@ subroutine score_general(p, t, start_index, filter_index, i_nuclide, &
+                i_nuc = mat % nuclide(l)
+
+                ! Get the current nuclide
+                nuc => nuclides(i_nuc)
```
I would just remove this and pass `nuclides(i_nuc)` to `nu_delayed` directly.

samuelshaner replied:
I'm getting the following compiler error when I make that switch:

    Error: Actual argument for 'nuc' must be a pointer at (1)

Is there a way to pass the pointer to the nuclide with the change you recommended?

paulromano replied:
The formal argument in `nu_delayed` doesn't need the `pointer` attribute and can just be declared `intent(in)`.

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1041,122 @@ subroutine score_fission_eout(p, t, i_score)
+    integer :: bin_energyout ! original outgoing energy bin
+    real(8) :: score         ! actual score
+    real(8) :: E_out         ! energy of fission bank site
+    logical :: d_found = .FALSE. ! bool to inidicate if delayed group was found
```
lowercase `.false.` is the most commonly used elsewhere in the code

paulromano commented on src/particle_header.F90:
```diff
@@ -102,6 +106,7 @@ module particle_header
   subroutine initialize_particle(this)
 
     class(Particle) :: this
+    integer         :: d
```
Remove this unused variable

paulromano commented on src/physics.F90:
```diff
@@ -1063,14 +1063,16 @@ subroutine create_fission_sites(p, i_nuclide, i_reaction)
-    real(8) :: phi          ! fission neutron azimuthal angle
-    real(8) :: weight       ! weight adjustment for ufs method
-    logical :: in_mesh      ! source site in ufs mesh?
+    integer :: d                        ! delayed group index
```
Unused variable

paulromano commented on src/ace.F90:
```diff
@@ -672,6 +681,7 @@ subroutine read_nu_data(nuc)
 
     else
       nuc % nu_d_type = NU_NONE
+      nuc % n_precursor = ZERO
```
Should be `nuc % n_precursor = 0`?

samuelshaner replied:
you are correct, `nuc % n_precursor` is an integer and should be set with `nuc % n_precursor = 0`.

paulromano commented on src/tracking.F90:
```diff
@@ -29,6 +29,7 @@ subroutine transport(p)
 
     type(Particle), intent(inout) :: p
 
+    integer :: d                      ! delayed group index
```
Unused variable

paulromano commented:
@samuelshaner Thanks for updating this PR and especially for fixing the performance issues. At least on the models I'm testing, I don't see any performance decrement due to the addition of delayed group filters and delayed nu fission scores. I think this one should be ready to go once the latest round of mostly trivial comments are resolved.

samuelshaner commented:
@paulromano thanks for the comments. I fixed the issues you noted.

I have also fixed an issue I had created dealing with the merging of tallies in the Python API. In short, tallies cannot be merged if only one of them contains a `delayedgroup` filter. Also, when merging the bins of a filter, the numpy ndarrays are now concatenated. Before they were either added element-wise (if the num bins was equal) or the code would error out (if the num bins was not equal).

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1032,122 @@ subroutine score_fission_eout(p, t, i_score)
+    integer :: bin_energyout ! original outgoing energy bin
+    real(8) :: score         ! actual score
+    real(8) :: E_out         ! energy of fission bank site
+    logical :: d_found = .false. ! bool to inidicate if delayed group was found
```
No need to initialize this variable. Also, keep in mind that when you initialize a variable where it is declared, the initialization is done only once at program startup (not every time the subroutine is entered).

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1032,122 @@ subroutine score_fission_eout(p, t, i_score)
+             E_out > t % filters(i) % real_bins(n)) cycle
+
+        ! check if delayed group is in delayed group bins
+        if (j > 0) then
```
Isn't `j > 0` by default? I.e. If there's no delayedgroup filter, this subroutine is never entered.

samuelshaner replied:
This subroutine will be entered when `delayed-nu-fission` is requested with an energy out filter. The delayed group filter (with index `j` in this subroutine) does not necessarily have to be present. If it is not present, the total `delayed-nu-fission` rate is tallied; If it is present; the `delayed-nu-fission` rate is tallied for the requested delayed group bins.

I will push changes shortly that modify this subroutine to clarify the procedure for the cases where a delayed group filter is present and where a delayed group filter is not present.

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1032,122 @@ subroutine score_fission_eout(p, t, i_score)
+          do d_bin = 1, t % filters(j) % n_bins
+            d = t % filters(j) % int_bins(d_bin)
+            if (d == g) then
+              d_found = .TRUE.
```
Use .true. rather than .TRUE.

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1032,122 @@ subroutine score_fission_eout(p, t, i_score)
+
+          ! if the delayedgroup filter is present and the delayed group is not
+          ! one of the delayedgroup bins, go to next particle in bank.
+          if (d_found .eqv. .FALSE.) cycle
```
`if (.not. d_found) cycle` is clearer.

paulromano commented on src/tally.F90:
```diff
@@ -845,6 +1032,122 @@ subroutine score_fission_eout(p, t, i_score)
+
+      ! get the delayed group
+      g = fission_bank(n_bank - p % n_bank + k) % delayed_group
+      d_found = .FALSE.
```
Use .false. instead of .FALSE.

paulromano commented:
@samuelshaner Thanks again for the contribution and for working through all the issues!
